README.md
# Elliptic curve cryptography (ECC) in Dart ![Pub](https://img.shields.io/pub/v/viz_dart_ecc.svg)
[![Build Status](https://travis-ci.com/VizTower/viz_dart_ecc.svg?branch=master)](https://travis-ci.com/VizTower/viz_dart_ecc)

[[DOC](https://pub.dev/documentation/viz_dart_ecc/latest/) | [issue tracker](https://github.com/VizTower/viz_dart_ecc/issues)]

viz_dart_ecc is cryptography library for simply data encoding and decoding 
using VIZ blockchain's ECC cryptography algorithms. 

Code for viz_dart_ecc was forked from [eosdart_ecc](https://github.com/primes-network/eosdart_ecc) 
but most of it was rewritten for VIZ blockchain.

## Usage

A simple usage example:

```dart
import 'package:viz_dart_ecc/viz_dart_ecc.dart';

void main() {
  VIZPrivateKey privateKey = VIZPrivateKey.fromString(
      '5J2XSYiA62K5s9vLsXXpj9CdoGmWUnohEWnVmg8aJb8D2TYvpbW');

  VIZPublicKey publicKey = privateKey.toPublicKey();
  VIZSignature signature = privateKey.signString('data');

  print('Pub key: ' + publicKey.toString());
  print('Signatured data: ' + signature.toString());
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker](https://github.com/VizTower/viz_dart_ecc/issues)

CHANGELOG.md
# 1.0.1+1
 - add test for VIZPublicKey.toBuffer();
 - modified example and readme
 - modified pubspec.yml
# 1.0.1

 - Update example directory

# 1.0.0

 - Rewrite [eosdart_ecc](https://github.com/primes-network/eosdart_ecc) for VIZ blockchain.

test/signature_test.dart
import 'package:viz_dart_ecc/viz_dart_ecc.dart';
import 'package:test/test.dart';

void main() {
  group('VIZ signature tests.', () {
    test('Construct VIZ signature from string', () {
      String sigStr =
          '2030fe8c87efe5524e8f555bdf049b33cef2d82f33a312f0e731d21f20d60844376ea1330d592b9d62d7d3766b80f1c40a71564c4bb7a9958bc74cf757d9e97429';
      VIZSignature signature = VIZSignature.fromString(sigStr);
      print(signature);

      expect(sigStr, signature.toString());
    });
  });
}


test/viz_dart_ecc_test.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:viz_dart_ecc/viz_dart_ecc.dart';
import 'package:test/test.dart';

void main() {
  group('VIZ Key tests.', () {
    test('Construct VIZ public key from string', () {
      VIZPublicKey publicKey = VIZPublicKey.fromString(
          'VIZ8VmBfvDi5S2Y8pB4T5AyfZmmqGBL2jV4t5H5cZbonp1VgSWh3u');
      print(publicKey);

      expect('VIZ8VmBfvDi5S2Y8pB4T5AyfZmmqGBL2jV4t5H5cZbonp1VgSWh3u',
          publicKey.toString());
    });

    test('Transform public key to buffer', () {
      VIZPublicKey publicKey = VIZPublicKey.fromString(
          'VIZ8VmBfvDi5S2Y8pB4T5AyfZmmqGBL2jV4t5H5cZbonp1VgSWh3u');
      print(publicKey);

      Uint8List buf = publicKey.toBuffer();

      VIZPublicKey newPubKey = VIZPublicKey.fromBuffer(buf);

      expect(newPubKey.toString(),
          'VIZ8VmBfvDi5S2Y8pB4T5AyfZmmqGBL2jV4t5H5cZbonp1VgSWh3u');
    });

    test('Construct VIZ private key from string', () {
      // common private key
      VIZPrivateKey privateKey = VIZPrivateKey.fromString(
          '5KWh6ThtRJfjzTJJ4EWkugrzLV1zVXZWSFPg3qTae2uTTq8pM9U');
      expect('VIZ8VmBfvDi5S2Y8pB4T5AyfZmmqGBL2jV4t5H5cZbonp1VgSWh3u',
          privateKey.toPublicKey().toString());
      expect('5KWh6ThtRJfjzTJJ4EWkugrzLV1zVXZWSFPg3qTae2uTTq8pM9U',
          privateKey.toString());
    });

    test('Invalid VIZ private key', () {
      try {
        VIZPrivateKey.fromString(
            '5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjsm');
        fail('Should be invalid private key');
      } on InvalidKey {} catch (e) {
        fail('Should throw InvalidKey exception');
      }
    });

    test('Construct random VIZ private key from seed', () {
      VIZPrivateKey privateKey = VIZPrivateKey.fromSeed('abc');
      print(privateKey);
      print(privateKey.toPublicKey());

      VIZPrivateKey privateKey2 =
          VIZPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toPublicKey().toString(),
          privateKey2.toPublicKey().toString());
    });

    test('Construct random VIZ private key', () {
      VIZPrivateKey privateKey = VIZPrivateKey.fromRandom();

      print(privateKey);
      print(privateKey.toPublicKey());

      VIZPrivateKey privateKey2 =
          VIZPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toPublicKey().toString(),
          privateKey2.toPublicKey().toString());
    });

    test('Sign transaction with VIZ private key', () {
      VIZPrivateKey privateKey = VIZPrivateKey.fromString(
          '5J2XSYiA62K5s9vLsXXpj9CdoGmWUnohEWnVmg8aJb8D2TYvpbW');
      VIZSignature signature = privateKey.sign(hex.decode('ff00ee65'));

      print(signature.toString());

      expect(signature.toString(),
          '1f02a25804ad71bd4226486f94dcad358426e5bde9a68eb69e5ed18552d81843df011c1d598c7d4451b6d0fe3c3e7f885c46b88577616895fcef80eb82ae9ba288');
    });
  });
}


lib/src/key.dart
import 'dart:typed_data';
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/src/utils.dart';
import 'package:pointycastle/ecc/api.dart' show ECSignature, ECPoint;

import './exception.dart';
import './key_base.dart';
import './signature.dart';

/// VIZ Public Key
class VIZPublicKey extends VIZKey {
  static final PUB_KEY_PREFIX = 'VIZ';

  ECPoint q;

  /// Construct VIZ public key from buffer
  VIZPublicKey.fromPoint(this.q);

  /// Construct VIZ public key from string
  factory VIZPublicKey.fromString(String keyStr) {
    if (!keyStr.startsWith(PUB_KEY_PREFIX)) {
      throw InvalidKey('A key must begin with "$PUB_KEY_PREFIX"');
    }
    String pupKeyStr = keyStr.substring(3);
    Uint8List keyBuf = VIZKey.decodeKey(pupKeyStr);
    return VIZPublicKey.fromBuffer(keyBuf);
  }

  factory VIZPublicKey.fromBuffer(Uint8List buffer) {
    ECPoint point = VIZKey.secp256k1.curve.decodePoint(buffer);
    return VIZPublicKey.fromPoint(point);
  }

  Uint8List toBuffer() {
    // always compressed
    return q.getEncoded(true);
  }

  String toString() {
    return PUB_KEY_PREFIX + VIZKey.encodeKey(this.toBuffer());
  }
}

/// VIZ Private Key
class VIZPrivateKey extends VIZKey {
  Uint8List d;

  BigInt _r;
  BigInt _s;

  /// Constructor VIZ private key from the key buffer itself
  VIZPrivateKey.fromBuffer(this.d);

  /// Construct the private key from WIF string
  VIZPrivateKey.fromString(String keyStr) {
    // WIF
    Uint8List keyBuf = VIZKey.decodeKey(keyStr, VIZKey.SHA256X2);
    int version = keyBuf.first;
    if (VIZKey.VERSION != version) {
      throw InvalidKey('Expected version ${0x80}, instead got ${version}');
    }

    d = keyBuf.sublist(1, keyBuf.length);
    if (d.lengthInBytes == 33 && d.elementAt(32) == 1) {
      // remove compression flag
      d = d.sublist(0, 32);
    }

    if (d.lengthInBytes != 32) {
      throw InvalidKey('Expecting 32 bytes, got ${d.length}');
    }
  }

  /// Generate VIZ private key from seed. Please note: This is not random!
  /// For the given seed, the generated key would always be the same
  factory VIZPrivateKey.fromSeed(String seed) {
    Digest s = sha256.convert(utf8.encode(seed));
    return VIZPrivateKey.fromBuffer(s.bytes);
  }

  /// Generate the random VIZ private key
  factory VIZPrivateKey.fromRandom() {
    final int randomLimit = 1 << 32;
    Random randomGenerator;
    try {
      randomGenerator = Random.secure();
    } catch (e) {
      randomGenerator = new Random();
    }

    int randomInt1 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy1 = encodeBigInt(BigInt.from(randomInt1));

    int randomInt2 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy2 = encodeBigInt(BigInt.from(randomInt2));

    int randomInt3 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy3 = encodeBigInt(BigInt.from(randomInt3));

    List<int> entropy = entropy1.toList();
    entropy.addAll(entropy2);
    entropy.addAll(entropy3);
    Uint8List randomKey = Uint8List.fromList(entropy);
    Digest d = sha256.convert(randomKey);
    return VIZPrivateKey.fromBuffer(d.bytes);
  }

  /// Check if a private key is WIF format
  static bool isWIF(String wif) {
    try {
      VIZKey.decodeKey(wif, VIZKey.SHA256X2);
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Get the public key string from this private key
  VIZPublicKey toPublicKey() {
    BigInt privateKeyNum = decodeBigInt(this.d);
    ECPoint ecPoint = VIZKey.secp256k1.G * privateKeyNum;

    return VIZPublicKey.fromPoint(ecPoint);
  }

  /// Sign the bytes data using the private key
  VIZSignature sign(Uint8List data) {
    Digest d = sha256.convert(data);
    return signHash(d.bytes);
  }

  /// Sign the string data using the private key
  VIZSignature signString(String data) {
    return sign(utf8.encode(data));
  }

  /// Sign the SHA256 hashed data using the private key
  VIZSignature signHash(Uint8List sha256Data) {
    int nonce = 0;
    BigInt n = VIZKey.secp256k1.n;
    BigInt e = decodeBigInt(sha256Data);

    while (true) {
      _deterministicGenerateK(sha256Data, this.d, e, nonce++);
      var N_OVER_TWO = n >> 1;
      if (_s.compareTo(N_OVER_TWO) > 0) {
        _s = n - _s;
      }
      ECSignature sig = ECSignature(_r, _s);

      Uint8List der = VIZSignature.ecSigToDER(sig);

      int lenR = der.elementAt(3);
      int lenS = der.elementAt(5 + lenR);
      if (lenR == 32 && lenS == 32) {
        int i = VIZSignature.calcPubKeyRecoveryParam(
            decodeBigInt(sha256Data), sig, this.toPublicKey());
        i += 4; // compressed
        i += 27; // compact  //  24 or 27 :( forcing odd-y 2nd key candidate)
        return VIZSignature(i, sig.r, sig.s);
      }
    }
  }

  String toString() {
    List<int> version = List<int>();
    version.add(VIZKey.VERSION);
    Uint8List keyBuf = VIZKey.concat(Uint8List.fromList(version), this.d);

    return VIZKey.encodeKey(keyBuf, VIZKey.SHA256X2);
  }

  BigInt _deterministicGenerateK(
      Uint8List hash, Uint8List x, BigInt e, int nonce) {
    List<int> newHash = hash;
    if (nonce > 0) {
      List<int> addition = Uint8List(nonce);
      List<int> data = List.from(hash)..addAll(addition);
      newHash = sha256.convert(data).bytes;
    }

    // Step B
    Uint8List v = Uint8List(32);
    for (int i = 0; i < v.lengthInBytes; i++) {
      v[i] = 1;
    }

    // Step C
    Uint8List k = Uint8List(32);

    // Step D
    List<int> d1 = List.from(v)
      ..add(0)
      ..addAll(x)
      ..addAll(newHash);

    Hmac hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    k = hMacSha256.convert(d1).bytes;

    // Step E
    hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    v = hMacSha256.convert(v).bytes;

    // Step F
    List<int> d2 = List.from(v)
      ..add(1)
      ..addAll(x)
      ..addAll(newHash);

    k = hMacSha256.convert(d2).bytes;

    // Step G
    hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
    v = hMacSha256.convert(v).bytes;
    // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
    // Step H2b again
    v = hMacSha256.convert(v).bytes;

    BigInt T = decodeBigInt(v);
    // Step H3, repeat until T is within the interval [1, n - 1]
    while (T.sign <= 0 ||
        T.compareTo(VIZKey.secp256k1.n) >= 0 ||
        !_checkSig(e, newHash, T)) {
      List<int> d3 = List.from(v)..add(0);
      k = hMacSha256.convert(d3).bytes;
      hMacSha256 = new Hmac(sha256, k); // HMAC-SHA256
      v = hMacSha256.convert(v).bytes;
      // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
      // Step H2b again
      v = hMacSha256.convert(v).bytes;

      T = decodeBigInt(v);
    }
    return T;
  }

  bool _checkSig(BigInt e, Uint8List hash, BigInt k) {
    BigInt n = VIZKey.secp256k1.n;
    ECPoint Q = VIZKey.secp256k1.G * k;

    if (Q.isInfinity) {
      return false;
    }

    _r = Q.x.toBigInteger() % n;
    if (_r.sign == 0) {
      return false;
    }

    _s = k.modInverse(VIZKey.secp256k1.n) * (e + decodeBigInt(d) * _r) % n;
    if (_s.sign == 0) {
      return false;
    }

    return true;
  }
}


lib/src/key_base.dart
import 'dart:typed_data';

import 'package:base58check/base58.dart';
import 'package:base58check/base58check.dart';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import 'package:pointycastle/src/utils.dart';
import "package:pointycastle/ecc/curves/secp256k1.dart";

import './exception.dart';

/// abstract VIZ Key
abstract class VIZKey {
  static final String SHA256X2 = 'sha256x2';
  static final int VERSION = 0x80;
  static final ECCurve_secp256k1 secp256k1 = ECCurve_secp256k1();

  static final base58 = Base58Codec(Base58CheckCodec.BITCOIN_ALPHABET);

  /// Decode key from string format
  static Uint8List decodeKey(String keyStr, [String keyType]) {
    Uint8List buffer = Uint8List.fromList(base58.decode(keyStr));

    Uint8List checksum = buffer.sublist(buffer.length - 4, buffer.length);
    Uint8List key = buffer.sublist(0, buffer.length - 4);

    Uint8List newChecksum;
    if (keyType == SHA256X2) {
      newChecksum = sha256x2(key).sublist(0, 4);
    } else {
      Uint8List check = key;
      newChecksum = RIPEMD160Digest().process(check).sublist(0, 4);
    }
    if (decodeBigInt(checksum) != decodeBigInt(newChecksum)) {
      throw InvalidKey("Invalid checksum: $checksum");
    }
    return key;
  }

  /// Encode key to string format using base58 encoding
  static String encodeKey(Uint8List key, [String keyType]) {
    if (keyType == SHA256X2) {
      Uint8List checksum = sha256x2(key).sublist(0, 4);
      return base58.encode(concat(key, checksum));
    }

    Uint8List keyBuffer = key;
    Uint8List checksum = RIPEMD160Digest().process(keyBuffer).sublist(0, 4);
    return base58.encode(concat(key, checksum));
  }

  /// Do SHA256 hash twice on the given data
  static Uint8List sha256x2(Uint8List data) {
    Digest d1 = sha256.convert(data);
    Digest d2 = sha256.convert(d1.bytes);
    return d2.bytes;
  }

  static Uint8List concat(Uint8List p1, Uint8List p2) {
    List<int> keyList = p1.toList();
    keyList.addAll(p2);
    return Uint8List.fromList(keyList);
  }

  static List<int> toSigned(Uint8List bytes) {
    List<int> result = List();
    for (int i = 0; i < bytes.length; i++) {
      int v = bytes[i].toSigned(8);
      //TODO I don't know why, just guess...
      if (i == 0 && v < 0) {
        result.add(0);
      }
      result.add(v);
    }
    return result;
  }
}


lib/src/signature.dart
import 'dart:typed_data';
import 'dart:convert';

import 'package:convert/convert.dart';
import 'package:crypto/crypto.dart';
import "package:pointycastle/api.dart" show PublicKeyParameter;
import 'package:pointycastle/ecc/api.dart'
    show ECPublicKey, ECSignature, ECPoint;
import "package:pointycastle/signers/ecdsa_signer.dart";
import 'package:pointycastle/macs/hmac.dart';
import "package:pointycastle/digests/sha256.dart";
import 'package:pointycastle/src/utils.dart';

import './exception.dart';
import './key.dart';
import './key_base.dart';

/// VIZ Signature
class VIZSignature extends VIZKey {
  int i;
  ECSignature ecSig;

  /// Default constructor from i, r, s
  VIZSignature(this.i, BigInt r, BigInt s) {
    this.ecSig = ECSignature(r, s);
  }

  /// Construct VIZ signature from buffer
  VIZSignature.fromBuffer(Uint8List buffer) {
    if (buffer.lengthInBytes != 65) {
      throw InvalidKey(
          'Invalid signature length, got: ${buffer.lengthInBytes}');
    }

    i = buffer.first;

    if (i - 27 != i - 27 & 7) {
      throw InvalidKey('Invalid signature parameter');
    }

    BigInt r = decodeBigInt(buffer.sublist(1, 33));
    BigInt s = decodeBigInt(buffer.sublist(33, 65));
    this.ecSig = ECSignature(r, s);
  }

  /// Construct VIZ signature from string
  factory VIZSignature.fromString(String signatureStr) {
    Uint8List key = hex.decode(signatureStr);
    return VIZSignature.fromBuffer(key);
  }

  /// Verify the signature of the string raw data
  bool verify(String rawData, VIZPublicKey publicKey) {
    Digest d = sha256.convert(utf8.encode(rawData));

    return verifyHash(d.bytes, publicKey);
  }

  /// Verify the signature from in SHA256 hashed raw data
  bool verifyHash(Uint8List sha256Data, VIZPublicKey publicKey) {
    ECPoint q = publicKey.q;
    final signer = ECDSASigner(null, HMac(SHA256Digest(), 64));
    signer.init(false, PublicKeyParameter(ECPublicKey(q, VIZKey.secp256k1)));

    return signer.verifySignature(sha256Data, this.ecSig);
  }

  String toString() {
    List<int> b = List();
    b.add(i);
    b.addAll(encodeBigInt(this.ecSig.r));
    b.addAll(encodeBigInt(this.ecSig.s));

    Uint8List buffer = Uint8List.fromList(b);
    return hex.encode(buffer);
  }

  /// ECSignature to DER format bytes
  static Uint8List ecSigToDER(ECSignature ecSig) {
    List<int> r = VIZKey.toSigned(encodeBigInt(ecSig.r));
    List<int> s = VIZKey.toSigned(encodeBigInt(ecSig.s));

    List<int> b = List();
    b.add(0x02);
    b.add(r.length);
    b.addAll(r);

    b.add(0x02);
    b.add(s.length);
    b.addAll(s);

    b.insert(0, b.length);
    b.insert(0, 0x30);

    return Uint8List.fromList(b);
  }

  /// Find the public key recovery factor
  static int calcPubKeyRecoveryParam(
      BigInt e, ECSignature ecSig, VIZPublicKey publicKey) {
    for (int i = 0; i < 4; i++) {
      ECPoint Qprime = recoverPubKey(e, ecSig, i);
      if (Qprime == publicKey.q) {
        return i;
      }
    }
    throw 'Unable to find valid recovery factor';
  }

  /// Recovery VIZ public key from ECSignature
  static ECPoint recoverPubKey(BigInt e, ECSignature ecSig, int i) {
    BigInt n = VIZKey.secp256k1.n;
    ECPoint G = VIZKey.secp256k1.G;

    BigInt r = ecSig.r;
    BigInt s = ecSig.s;

    // A set LSB signifies that the y-coordinate is odd
    int isYOdd = i & 1;

    // The more significant bit specifies whether we should use the
    // first or second candidate key.
    int isSecondKey = i >> 1;

    // 1.1 Let x = r + jn
    BigInt x = isSecondKey > 0 ? r + n : r;
    ECPoint R = VIZKey.secp256k1.curve.decompressPoint(isYOdd, x);
    ECPoint nR = R * n;
    if (!nR.isInfinity) {
      throw 'nR is not a valid curve point';
    }

    BigInt eNeg = (-e) % n;
    BigInt rInv = r.modInverse(n);

    ECPoint Q = (R * s + G * eNeg) * rInv;
    return Q;
  }
}
